pypiserver/pypiserver_app.py

```diff
@@ -0,0 +1,418 @@
+from collections import namedtuple
+from io import BytesIO
+from urllib.parse import urljoin, urlparse
+import functools
+import logging
+import mimetypes
+import os
+import re
+import xml.dom.minidom
+import xmlrpc.client as xmlrpclib  # py3
+import zipfile
+
+from pypiserver.config import RunConfig
+from pypiserver import __version__, core
+from pypiserver.bottle import (
+    static_file,
+    redirect,
+    request,
+    response,
+    HTTPError,
+    HTTPResponse,
+    Bottle,
+    template,
+)
+
+
+log = logging.getLogger(__name__)
+
+Upload = namedtuple("Upload", "pkg sig")
+
+_bottle_upload_filename_re = re.compile(r"^[a-z0-9_.!+-]+$", re.I)
+
+
+class PypiserverApp:
+    """A wrapper around the Pypiserver application."""
+
+    def __init__(self, config: RunConfig, app: Bottle) -> None:
+        """Construct a new Pypiserver application."""
+        self.app = app
+        self.config = config
+        self.auth = functools.partial(auth, config)
+        self._initialize_app()
+        # Add a reference to our config onto the bottle instance,
+        # so anyone using the app() constructor can get it if needed
+        self.app._pypiserver_config = config
```

Comment:
Hey, this looks great all of it, but if I may propose, how about making the factory and the wrapper class separate? Curious for arguments for and against. Then that would take the `_initialize_app()` method outside of this class. 

I think it could make it easier like this for example: 

```python
class PypiserverApp:
    def __init__(self, config: RunConfig) -> None:
        # without any reference to Bottle here
        self.config = config
        self.auth = functools.partial(auth, config)
   
     def log_request(self) -> None:
        log.info(self.config.log_req_frmt, request.environ)
     
     # and so forth with the other hook methods...

class PypiserverAppFactory:

    @classmethod
    def build_app(cls, config: RunConfig, app: Bottle) -> Bottle:
        pypiserver_app = PypiserverApp(config)
        return cls._initialize_app(app, pypiserver_app)

   @staticmethod
    def _initialize_app(app: Bottle, pypiserver_app) -> Bottle:
        app.hook("before_request")(pypiserver_app.log_request)
        app.hook("before_request")(pypiserver_app.parse_request_path)
        # and so on...
        app._pypiserver_config = config
        return app
```
Then potentially, later this would allow extending the `pypiserver` with ohther backends for people who could want it, if instead of `app: Bottle` the parameter will be `builder: AppBuilder` or so that will be an instance of a strategy to initialize an app of a particular backend given the pypiserver wrapper. But that's a bit too far off maybe. Let me know :)

Reply:
Yeah, I like this idea. I'll take a look at updating the PR sometime soon!